### PR TITLE
Implement buffering and caching

### PR DESCRIPTION
**Pull request by somewhatabstract**
# Summary

This adds methods to support the different ways we may want to resolve requests. This is an evolution of the `react-render-server` code, with the aim of improving versatility and simplifying usage.

The main differences are:

1. The ability to tell if a response is from the cache is encapsulated in the request operation
2. Buffering is optional (though we're likely to always want it - it gives us an option)

I tested out the new `fromCache` behavior in a local modification of the `react-render-server` to verify that it will work. It would be nice to have some integration level tests for some of this once all the bits are in place. That will likely happen later in the development of `render-gateway` and be based on our integration tests for `react-render-server`.

## Issues

Part of WEB-1858

## Review Guidelines

In this version, the superagent-cache-plugin is recreated for every request. This is different from `react-render-server` which reuses the same instance across every request. Just as with the agentkeepalive instance that is currently remade every request, this will be refactored in an upcoming change to allow requests to share these instances. I did not do that right away as I wanted to keep the diffs an evolution of the code so that it was easy to follow along.

I also felt that delaying this refactor would help to ensure that the code made sense. Without some use in place, it is perhaps harder to see where the responsibility of creating these shared pieces should lie.

### CODE REVIEW

File: src/requests-from-cache.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+    SuperAgentRequest,
+    Response as SuperAgentResponse,
+} from "superagent";
+import type {CachingStrategy} from "./types.js";
+
+/**
+ * This is the name of the property we attach to responses so that we can
+ * indicate if a response was from the cache or not.
+ */
+export const FROM_CACHE_PROP_NAME: string = "_fromCache";
```
Comment (somewhatabstract) [inline src/requests-from-cache.js]: Exported so that our tests don't have to know what this string really is.

File: src/requests-from-cache.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+             *
+             * It also works because the response we get here is what is in
+             * the cache so any modifications we make are reflected in the
+             * cached value. Cheeky, but it works 😈
+             */
+            res[FROM_CACHE_PROP_NAME] =
+                res[FROM_CACHE_PROP_NAME] !== NEW_CACHE_ENTRY;
+            return res;
+        });
+};
```
Comment (somewhatabstract) [inline src/requests-from-cache.js]: I will reiterate here that I tested this in the `react-render-server` locally to check that it worked as expected.

There is no test in this code yet that fully validates this works. We will eventually want some integration tests to verify things like this, but I didn't want to get into building those while code is still being erected into its final form.

**Comment by somewhatabstract** [comment]
> Looks ok to me! Thank you for adding so many tests.

I would love to have some additional coverage beyond just unit tests, but this at least gives us some level of confidence on what is expected to occur. Hopefully, I can write some code that uses all this and then we can actually get some sort of soak testing going with `ab` and `parallel` to check out cache perf with different cache setups :)

File: src/__tests__/requests-from-cache.test.js [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+            asCachedRequest(fakeRequest, fakeCachingStrategy, fakeBufferArg);
+            const pruneFn = fakeRequest.prune.mock.calls[0][0];
+
+            // Act
+            pruneFn(fakeResponse, gutResponseFn);
+
+            // Assert
+            expect(fakeResponse).toHaveProperty(FROM_CACHE_PROP_NAME);
+            expect(fakeResponse[FROM_CACHE_PROP_NAME]).not.toBeBoolean();
+            expect(fakeResponse[FROM_CACHE_PROP_NAME]).not.toBeFalsy();
```
[the remark this answers is not in the document]
Comment (somewhatabstract) [reply src/__tests__/requests-from-cache.test.js]: No, not at all. This is a convenience thing we add to track if something comes from cache. It initializes to NO and then we use that to determine if we should set it to true or false. :)

[the remark this answers is not in the document]
Comment (somewhatabstract) [reply src/__tests__/requests-from-cache.test.js]: So, when we first finish the request, it was just cached and so it reads "NO". In that case, we set it to `false` as in "false, it was not from the cache". On subsequent calls, that cached version will be the response and we'll see it set to `false` and therefore set it to `true` as in "true, this was from the cache".

There are comments in the code about this and I tested it all out in the `react-render-server` to check it works. The `react-render-server` has a version of this but it's a very leaky abstraction there and I wanted to encapsulate it more effectively in this version.

File: src/requests-from-cache.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+ *
+ * @param {SuperAgentResponse} response The response to check.
+ * @returns {boolean} true if the response was from cache; otherwise false.
+ */
+export const isFromCache = (response: SuperAgentResponse): boolean =>
+    response[FROM_CACHE_PROP_NAME] === true;
+
+/**
+ * Turn unbuffered, uncached request into uncached request with or without
+ * buffer.
```
[the remark this answers is not in the document]
Comment (somewhatabstract) [reply src/requests-from-cache.js]: Good question. It means exactly what it says. :)

We're taking a raw request that doesn't support our "from cache" flag and turning it into one that does, and will have that flag set to "false". We're also adding the buffer call.

File: src/requests-from-cache.js [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+ */
+export const asCachedRequest = (
+    request: SuperAgentRequest,
+    strategy: CachingStrategy,
+    buffer: boolean,
+): SuperAgentRequest => {
+    const {provider, getExpiration} = strategy;
+
+    const superagentCache = superagentCachePlugin(provider);
+    const NEW_CACHE_ENTRY = "NO";
```
Comment (somewhatabstract) [inline src/requests-from-cache.js]: Perhaps this constant needs a better name.

File: src/__tests__/requests-from-cache.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply src/__tests__/requests-from-cache.test.js]: > So, when we first finish the request, it was just cached and so it reads "NO". In that case, we set it to false as in "false, it was not from the cache". On subsequent calls, that cached version will be the response and we'll see it set to false and therefore set it to true as in "true, this was from the cache".

I had some trouble grok the comments, but this was really helpful.

File: src/requests-from-cache.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (kevinbarabash) [reply src/requests-from-cache.js]: The word "raw" helps a lot here.  For some reason I thought maybe we'd want to convert a cached request back to an uncached request so I was thinking of `asUncachedRequest` as being the reverse of `asCachedRequest` and that composing the two functions would result in the identity function.  That is not the case though.

[the remark this answers is not in the document]
Comment (somewhatabstract) [reply src/requests-from-cache.js]: Yeah, naming things is hard. I could use opaque types or aliased types to make it clearer

File: src/__tests__/requests-from-cache.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (somewhatabstract) [reply src/__tests__/requests-from-cache.test.js]: OK, I have renamed the `NEW_CACHE_ENTRY` constant to `FRESHLY_PRUNED` and updated the commentary to try and make it clearer how it all works. I'm going to merge it, but if you have further comments, LMK and I can address them in a future PR.

File: src/requests-from-cache.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (somewhatabstract) [reply src/requests-from-cache.js]: I'm going to leave this as-is here, but consider changing things a bit in the diff that uses all this stuff.

**Pull request merged**